### PR TITLE
convert external post processor to use new DataObjects

### PR DESCRIPTION
**Pull request by wangcj05**
--------
Pull Request Description
--------
##### What are the significant changes in functionality due to this change request?

----------------
For Change Control Board: Change Request Review
----------------
The following review must be completed by an authorized member of the Change Control Board.
- [ ] 1. Review all computer code.
- [ ] 2. If any changes occur to the input syntax, there must be an accompanying change to the user manual and xsd schema. If the input syntax change deprecates existing input files, a conversion script needs to be added (see Conversion Scripts).
- [ ] 3. Make sure the Python code and commenting standards are respected (camelBack, etc.) - See on the [wiki](https://github.com/idaholab/raven/wiki/RAVEN-Code-Standards#python) for details.
- [ ] 4. Automated Tests should pass, including run_tests, pylint, manual building and xsd tests. If there are changes to Simulation.py or JobHandler.py the qsub tests must pass.
- [ ] 5. If significant functionality is added, there must  be tests added to check this. Tests should cover all possible options.  Multiple short tests are preferred over one large test. If new development on the internal JobHandler parallel system is performed, a cluster test must be added setting, in <RunInfo> XML block, the node ```<internalParallel>``` to True.
- [ ] 6. If the change modifies or adds a requirement or a requirement based test case, the Change Control Board's Chair or designee also needs to approve the change.  The requirements and the requirements test shall be in sync.
- [ ] 7. The merge request must reference an issue.  If the issue is closed, the issue close checklist shall be done.
- [ ] 8. If an analytic test is changed/added is the the analytic documentation updated/added?

### CODE REVIEW

**Comment by wangcj05** [comment]
1. This PR will break the unit tests for the DataObjects.
2. I think the following test can be removed:
 ```tests/framework/PostProcessors/ExternalPostProcessor/test_externalPostProcessor.xml```

**Comment by wangcj05** [comment]
DataObjects have been fixed, and this PR is ready for review. @alfoa @PaulTalbot-INL

**Comment by wangcj05** [comment]
@alfoa I do make a lot of changes in this External PP, Please take a look at this PR.

File: framework/DataObjects/XDataSet.py [hunk 1]
```diff
@@ -938,8 +940,6 @@ def _fromCSV(self,fName,**kwargs):
       @ In, kwargs, dict, optional arguments
       @ Out, None
     """
-    assert(self._data is None)
-    assert(self._collector is None)
```
Comment (alfoa) [inline framework/DataObjects/XDataSet.py]: did you need to remove this?

File: framework/DataObjects/XDataSet.py [hunk 2]
```diff
@@ -1032,8 +1032,6 @@ def _fromDict(self,source,dims=None,**kwargs):
       @ In, kwargs, dict, optional, additional arguments
       @ Out, None
     """
-    assert(self._data is None)
-    assert(self._collector is None)
```
Comment (alfoa) [inline framework/DataObjects/XDataSet.py]: same as above?

File: framework/PostProcessors/ExternalPostProcessor.py [hunk 3]
```diff
@@ -90,52 +91,44 @@ def inputToInternal(self, currentInput):
...
+    if type(currentInput) == list:
+      if len(currentInput) != 1:
+        self.raiseAnError(IOError, "The postprocessor ", self.name, "only allows one input DataObjects,"
+              + " but multiple inputs are provided!")
+      else:
+        currentInput = currentInput[-1]
+    if hasattr(currentInput, 'type'):
+      inType = currentInput.type
+    else:
+      self.raiseAnError(IOError, "Input type ", type(currentInput).__name__, ' is not recognized!')
```
Comment (alfoa) [inline framework/PostProcessors/ExternalPostProcessor.py]: In raven, all the objects have the ```.type``` attribute.

File: framework/PostProcessors/ExternalPostProcessor.py [hunk 4]
```diff
@@ -332,42 +223,55 @@ def run(self, inputIn):
...
-    for target in input['targets'].keys():
-      if target not in outputDict.keys() and target in input['targets'].keys():
-        outputDict[target] = input['targets'][target]
+    # TODO: We assume the structure of input to the external pp isthe  same as the struture of output to this external pp
+    # An interface pp should be used if the user wants to merge two data objects, or change the structures of input data
+    # objects.
+    numRlz = len(outputDict.values()[0])
+    for val in outputDict.values():
+      if len(val) != numRlz:
+        self.raiseAnError(IOError, "The return results from the external functions have different number of lengths!"
```
Comment (alfoa) [inline framework/PostProcessors/ExternalPostProcessor.py]: ```different number of lengths -> different number of realizations ```

File: framework/DataObjects/XDataSet.py (hunk 2 above)
Comment (wangcj05) [reply framework/DataObjects/XDataSet.py]: the check is moved to the load method

File: framework/PostProcessors/ExternalPostProcessor.py (hunk 3 above)
Comment (wangcj05) [reply framework/PostProcessors/ExternalPostProcessor.py]: Is it true? we used to check this.

Comment (wangcj05) [reply framework/PostProcessors/ExternalPostProcessor.py]: fixed

File: framework/PostProcessors/ExternalPostProcessor.py (hunk 4 above)
Comment (wangcj05) [reply framework/PostProcessors/ExternalPostProcessor.py]: fixed

**Pull request merged**